src/Microsoft.TestPlatform.Utilities/CommandLineUtilities.cs

```diff
@@ -33,49 +33,73 @@ public static bool SplitCommandLineIntoArguments(string args, out string[] argum
                 if (args[index] == '#')
                 {
                     index++;
-                    while (args[index] != '\n')
+                    while (index < args.Length && args[index] != '\n')
                     {
                         index++;
                     }
+
+                    // We are done processing comment move to next statement.
                     continue;
                 }
 
-                // do one argument
+                // Read argument until next whitespace (not in quotes).
                 do
                 {
                     if (args[index] == '\\')
                     {
-                        int cSlashes = 1;
+                        // Move to next char.
                         index++;
-                        while (index == args.Length && args[index] == '\\')
-                        {
-                            cSlashes++;
-                        }
 
-                        if (index == args.Length || args[index] != '"')
+                        // If this was the last char then output the slash.
+                        if (index == args.Length)
                         {
-                            currentArg.Append('\\', cSlashes);
+                            currentArg.Append('\\');
+
+                            index++;
+                            continue;
                         }
                         else
                         {
-                            currentArg.Append('\\', (cSlashes >> 1));
-                            if (0 != (cSlashes & 1))
+                            // If the char after '\' is also a '\', output the second '\' and skip over to the next char.
+                            if (args[index] == '\\')
+                            {
+                                currentArg.Append('\\');
+
+                                // We processed the escaped \, move to next char.
+                                index++;
+                                continue;
+                            }
+
+                            // If the char after '\' is a '"', output '"' and skip over to the next char.
+                            if (index <= args.Length && args[index] == '"')
                             {
                                 currentArg.Append('"');
```

Comment:
Main bug was here imho. We handled the quote, but we don't advance to the next character because the logic is trying to be too clever with the `\` being or not being doubled.

The " is then processed in the next loop, and stops the string from being quoted, even though the quote was escaped.